 **Archivo Ejecución**</br>
 Francisco Nortes Parra</br>
 Lamine Mouhamadou </br>
 Ariana Escobar Ros

**CASO DE NEGOCIO** </br>
HeadHunter es una empresa líder en España especializada en la captación de talento. Sin embargo, su enfoque tradicional representa un desafío en un mercado en constante evolución. Para garantizar su posición de liderazgo, es fundamental que la empresa se adapte a las nuevas tendencias y exigencias del sector.
El objetivo principal de este análisis es identificar y comprender las tendencias del mercado, permitiendo así que HeadHunter modernice sus estrategias y continúe siendo un referente en la industria del reclutamiento.



**METODOLOGÍA SEGUIDA** </br>

A continuación veréis toda la metodología seguida, diferenciada por apartados para su mayor claridad y con la distribución de tiempo que hemos utilizado en cada parte. El programa utilizado para esta parte del análisis ha sido Google Colab y fue elegido por su capacidad de colaboración en tiempo real y por la facilidad con la que permite manejar grandes volúmenes de datos.


## Importacion de Librerias y lector de archivos.


Importamos las librerias en este caso de Pandas y el lector de tablas Drive.

In [ ]:
import pandas as pd

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


##Limpieza de tablas


### **01 - Tabla GlassDoor**

Una vez importados los datos con pandas cargamos las 3 tablas de datos que teníamos principalmente y cada uno de nosotros empezó a trabajar en cada una de ellas.

In [ ]:
# Leer el archivo CSV
data3 = pd.read_csv('/content/drive/MyDrive/HEAD HUNTERS/Data Sets (NO BORRAR NI EXPORTAR)/dataset_glassdoor-jobs-scraper_2023-02-05_01-07-13-866 (1).csv')

# Seleccionar las columnas requeridas
required_columns = ['jobTitle', 'jobDetails', 'jobLocation', 'employerName', 'salary/min', 'salary/max', 'salary/currency', 'datePosted']
data_cleaned = data3[required_columns]

# Renombrar las columnas
new_column_names = {
    'jobTitle': 'Puesto',
    'jobDetails': 'Descripcion',
    'jobLocation': 'Ciudad',
    'employerName': 'Nombre_Empresa',
    'salary/min': 'Salario_Minimo',
    'salary/max': 'Salario_Maximo',
    'salary/currency': 'Moneda',
    'datePosted': 'Fecha_Publicacion'
}

data_cleaned.rename(columns=new_column_names, inplace = True)

# Reemplazar valores NaN con 0
data_cleaned.fillna(0, inplace = False)

# Convertir solo las columnas de tipo float a strings
for column in data_cleaned.select_dtypes(include=['float']).columns:
    data_cleaned[column] = data_cleaned[column]

# Convertir todo el DataFrame a string
tabla1 = data_cleaned

# Ordenar las columnas según el orden especificado
orden_columnas = [
    'Puesto',
    'Descripcion',
    'Nombre_Empresa',
    'Ciudad',
    'Salario_Minimo',
    'Salario_Maximo',
    'Moneda',
    'Fecha_Publicacion'
]

# Acceder a las columnas ordenadas
tabla1_1 = tabla1[orden_columnas]

print(tabla1_1.dtypes)  # Para verificar que todas las columnas sean strings
tabla1_1

Puesto                object
Descripcion           object
Nombre_Empresa        object
Ciudad                object
Salario_Minimo       float64
Salario_Maximo       float64
Moneda                object
Fecha_Publicacion     object
dtype: object


<ipython-input-5-b2cac1be1d24>:20: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_cleaned.rename(columns=new_column_names, inplace = True)
<ipython-input-5-b2cac1be1d24>:27: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_cleaned[column] = data_cleaned[column]


Puesto  \
0              Client Data Analytics & Models Analyst   
1             Pricing Data Analytics & Models Analyst   
2                                        Data Analyst   
3                            DATA SCIENTIST ANALYST I   
4                                Data Analyst Trainee   
5                                        Data Analyst   
6              Analyst - Business Strategy Consulting   
7                                        Data Analyst   
8                             Operations Data Analyst   
9       Portfolio & Strategy Analyst (Madrid or Nice)   
10  Data Analytics Analyst, SAS Programming - Euro...   
11       Analyst/Associate, Global Portfolio Advisory   
12                                Analyst, Valuations   
13                      PRÁCTICAS DATA ANALYST MADRID   
14                Data Management Analyst BNP PARIBAS   
15          Data Analyst BNP Paribas Personal Finance   
16                                    HR Data Analyst   
17                                       Data Analyst   
18  Oliver Wyman Digital: Data & Analytics - Analy...   
19                 Financial Reporting Analyst Intern   
20                                       Data Analyst   
21                                       Data Analyst   
22  Internship - Junior Social Media Analyst - Madrid   
23                     International Business Analyst   
24                            Data Analyst Specialist   
25     Graduate Program INTERNATIONAL FINANCE ANALYST   
26                             Analyst - GeoAnalytics   
27                             Data Labelling (Spain)   
28                                       Data Analyst   
29  Business Analyst - Private Equity & Principal ...   

                                          Descripcion  \
0   Client Data Analytics & Models Analyst\nCountr...   
1   Pricing Data Analytics & Models Analyst\nCount...   
2   Job Description\nEn Logicalis Spain queremos i...   
3   Area:\nCORPORATE & INVESTMENT BANKING\nCompany...   
4   En Alkemy España ¡Seguimos creciendo! Buscamos...   
5   Siemens Mobility is a separately managed compa...   
6   With approximately 85,000 employees in more th...   
7   Requisitos:Data Analyst, para desarrollar carr...   
8   Hi, we're Fever\nWe're excited you are checkin...   
9   Job Title\nPortfolio & Strategy Analyst (Madri...   
10  Join us on our exciting journey!\nPosition Pur...   
11  Description\nAnalyst/Associate, Global Portfol...   
12  Description\nAnalyst, Valuation Services\nAlva...   
13  IAG GBS LIMITED SUCURSAL EN ESPAÑA\nMadrid (Ma...   
14  The Group Chief Data Officer’s mission is arti...   
15  BNP Paribas Personal Finance, a través de su m...   
16  Una de las ventajas de trabajar en UNIR, es la...   
17  GroupM incorpora un Analista de datos con clar...   
18  WHO WE ARE\nOliver Wyman is a global leader in...   
19  Financial Reporting Analyst Intern\n\nMain res...   
20  Scalefast helps amazing brands develop a succe...   
21  ¿Qué es lo que estamos buscando?\nActualmente ...   
22  ¿Quienes somos?\n\n\nDigimind es el líder mund...   
23  Country / Region : Spain\nCity : Madrid\nBusin...   
24  Data Analyst Specialist\nFundada en 2008, Hipo...   
25  Agency :\nHavas Media Group Spain\nJob Descrip...   
26  WHAT YOU'LL DO\nGAMMA is BCG’s premier data sc...   
27  We are a leading digital identity company that...   
28  About Us & Our Culture\nWe, employees, need a ...   
29  Who You'll Work With\nYou will be based in our...   

                        Nombre_Empresa              Ciudad  Salario_Minimo  \
0                            Santander  Boadilla del Monte             NaN   
1                            Santander  Boadilla del Monte             NaN   
2                            Logicalis              Madrid             NaN   
3                                 BBVA              Madrid             NaN   
4                               Alkemy               Spain             NaN   
5                 Siemens Healthineers     

###**02 - Ofertas de empleo sample**

In [ ]:
import pandas as pd
df1 = pd.read_csv('/content/drive/MyDrive/HEAD HUNTERS/Data Sets (NO BORRAR NI EXPORTAR)/ofertas-de-empleo-sample (1).csv')

columnas_a_mantener = ['title', 'description', 'city', 'company_name', 'salary_min', 'salary_max', 'salary_currency', 'publish_date']
df_final = df1[columnas_a_mantener]

nuevo_nombre_columnas = {
    'title': 'Puesto',
    'description': 'Descripcion',
    'city': 'Ciudad',
    'company_name': 'Nombre_Empresa',
    'salary_min': 'Salario_Minimo',
    'salary_max': 'Salario_Maximo',
    'salary_currency': 'Moneda',
    'publish_date': 'Fecha_Publicacion'}

ultimate_df = df_final.rename(columns=nuevo_nombre_columnas)

orden_columnas = [
    'Puesto',
    'Descripcion',
    'Nombre_Empresa',
    'Ciudad',
    'Salario_Minimo',
    'Salario_Maximo',
    'Moneda',
    'Fecha_Publicacion'
]

df_ultimo = ultimate_df[orden_columnas]

df_ultimo.fillna(0, inplace = False)

tabla2 = df_ultimo[orden_columnas]
print(tabla2.dtypes)
tabla2

Puesto                object
Descripcion           object
Nombre_Empresa        object
Ciudad                object
Salario_Minimo       float64
Salario_Maximo       float64
Moneda                object
Fecha_Publicacion     object
dtype: object


Puesto  \
0                      Mecánico/a de Camión -REF 21217   
1                     Peón/a fin de semana - REF 20085   
2    Atención al público en sala VIP en el aeropuer...   
3    TECNICO/A SUPERIOR de Seguridad de la Informac...   
4                                    Administrativo/va   
..                                                 ...   
395                                       JEFE DE OBRA   
396      Comercial Telefónico / Prospección de Mercado   
397             ADMINISTRATIVO DEPARTAMENTO DE COMPRAS   
398                          Beca PRL y Medio Ambiente   
399                JUNIOR IT - SISTEMAS DE INFORMACION   

                                           Descripcion  \
0    Funciones\nSe busca una persona  responsable p...   
1    Buscamos colaboración en las siguientes tareas...   
2    Clece es una compañía líder a nivel nacional e...   
3    TECNICO/A SUPERIOR Seguridad de la Información...   
4    Somos una organización de más de 30 años de ex...   
..                                                 ...   
395  Empresa constructora especialista en obras del...   
396  ACEM, SL somos una empresa especializada en la...   
397  AVALCO es una central de compras de materiales...   
398  Desde Exide Technologies, empresa internaciona...   
399  Técnico de sistemas con ganas de aprender y ge...   

                                      Nombre_Empresa                   Ciudad  \
0                                 FCC Medio Ambiente                 Zaragoza   
1                                                FCC                 Zaragoza   
2                      Multiservicios Aeroportuarios                 Arrecife   
3                                             ENDESA                   Madrid   
4                                         Vilar Riba                      Vic   
..                                               ...                      ...   
395          Desarrollo de Tecnologías de Depuración  Carrión de los Céspedes   
396  AUXILIAR DE CARROCERIAS Y ESTRUCTURAS METALICAS    Camarma de Esteruelas   
397                                     GRUPO AVALCO                 Valencia   
398                               Exide Technologies        Alcalá de Henares   
399                         COMPANYIA GENERAL CARNIA                Barcelona   

     Salario_Minimo  Salario_Maximo Moneda Fecha_Publicacion  
0               NaN             NaN    NaN        2022-01-27  
1               NaN             NaN    NaN        2022-01-27  
2               NaN             NaN    NaN        2022-01-27  
3               NaN             NaN    NaN        2022-01-27  
4            8000.0         12000.0    EUR        2022-01-27  
..              ...             ...    ...               ...  
395         27000.0         33000.0    EUR        2022-01-25  
396             NaN             NaN    NaN        2022-01-25  
397          1200.0          1200.0    EUR        2022-01-25  
398          6000.0          8000.0    EUR        2022-01-25  
399         20000.0             NaN    EUR        2022-01-25  

[400 rows x 8 columns]

###**03 - ofertas de empleo**

In [ ]:
from pickle import FALSE
import pandas as pd
df2 = pd.read_csv('/content/drive/MyDrive/HEAD HUNTERS/Data Sets (NO BORRAR NI EXPORTAR)/ofertas-empleo (1).csv')

# Reemplazar los valores nulos por 0
df2.fillna(0, inplace = False)

#Renombrar el titulo de las columnas
df2.rename(columns={'title': 'Puesto', 'description': 'Descripcion', 'company_name': 'Nombre_Empresa', 'city':'Ciudad', 'salary_min':'Salario_Minimo', 'salary_max':'Salario_Maximo', 'salary_currency':'Moneda', 'publish_date':'Fecha_Publicacion'}, inplace=True)

# Eliminar las columans restantes

columnas_a_eliminar = ['contract_type', 'workday', 'salary_period', 'remote', 'country_code', 'salary_type', 'job_category']
df2a = df2.drop(columns=columnas_a_eliminar, axis =1)

# Sabemos los typos de datos
print(df2a.dtypes)

#Nos aseguramos que no haya ningun NaN
print(df2a[['Salario_Minimo', 'Salario_Maximo']])



print(df2a.dtypes)


tabla3 = df2a[orden_columnas]

tabla3

Puesto                object
Descripcion           object
Nombre_Empresa        object
Ciudad                object
Salario_Minimo       float64
Salario_Maximo       float64
Moneda                object
Fecha_Publicacion     object
dtype: object
      Salario_Minimo  Salario_Maximo
0                NaN             NaN
1            40000.0         45000.0
2            27000.0         30000.0
3                NaN             NaN
4            24000.0         27000.0
...              ...             ...
9995         85514.0        118912.0
9996         96000.0         96000.0
9997         77058.0        122825.0
9998         55247.0         70736.0
9999         30000.0         36000.0

[10000 rows x 2 columns]
Puesto                object
Descripcion           object
Nombre_Empresa        object
Ciudad                object
Salario_Minimo       float64
Salario_Maximo       float64
Moneda                object
Fecha_Publicacion     object
dtype: object


Puesto  \
0                         Técnico Gestión de la Demanda   
1     Java - SpringBoot - Proyecto estable - Teletra...   
2               Consultor SAP Hana Ingles Alto (Remoto)   
3           ANDROID DEVELOPER. INGLÉS ALTO. 100% REMOTO   
4                           Analista Funcional TEMPORAL   
...                                                 ...   
9995                                  Software Engineer   
9996                             Sommelier, Atlanta, GA   
9997                                         Devops USA   
9998                                    Math Designer I   
9999         Técnico/a de mantenimiento electromecánico   

                                            Descripcion  \
0     -Definición de roles y escenarios\r\n      - D...   
1     Desde AgioGlobal Technology, nos encontramos e...   
2     AirON compañia de innovacion servicios tecnolo...   
3     Desde MODIS IT consultora GLOBAL especializada...   
4     Nuevas aplicaciones/Extensión de la funcionali...   
...                                                 ...   
9995  Write clean, performant code to develop functi...   
9996  Be your own boss and set your own schedule. A ...   
9997  As a driving company in the SecDevOps field we...   
9998  Excellent spoken and written communication as ...   
9999  Mantenimiento preventivo y correctivo de maqui...   

                  Nombre_Empresa                  Ciudad  Salario_Minimo  \
0     Atento España - Estructura                  Madrid             NaN   
1         AGIO GLOBAL TECHNOLOGY                  Madrid         40000.0   
2            AIRON Sistemas S.L.                  Madrid         27000.0   
3                          MODIS                  Madrid             NaN   
4     Atento España - Estructura                  Madrid         24000.0   
...                          ...                     ...             ...   
9995                 Mobilewalla             Atlanta, GA         85514.0   
9996                    Cozymeal             Atlanta, GA         96000.0   
9997                   Iriusrisk             Atlanta, GA         77058.0   
9998     American Gaming Systems             Atlanta, GA         55247.0   
9999              Robert Walters  Vilafranca del Penedès         30000.0   

      Salario_Maximo Moneda Fecha_Publicacion  
0                NaN    NaN        2022-07-12  
1            45000.0    EUR        2022-07-12  
2            30000.0    EUR        2022-07-12  
3                NaN    NaN        2022-07-12  
4            27000.0    EUR        2022-07-12  
...              ...    ...               ...  
9995        118912.0    USD        2022-08-29  
9996         96000.0    USD        2022-08-29  
9997        122825.0    USD        2022-08-29  
9998         70736.0    USD        2022-08-29  
9999         36000.0    EUR        2022-03-28  

[10000 rows x 8 columns]

Cada uno utilizó sus códigos según su mayor comodidad con ellos pero teniendo todos los mismos objetivos:</br>

*   Eliminar las columans que no fueran las que acordamos previamente.
*   Ordenar las columnas de la manera acordada previamente
*   Eliminar nulos



Nos encontramos unos cuantos problemas en cuanto a la eliminación de nulos y en la definición de requerimientos que tenían que tener las tres tablas para que el análisis fuera óptimo. En general este paso se nos alargó por el hecho de la adaptación con el programa Google Colab. </br>





# HeadHunter

Una vez las columnas tenían la misma estructura, creamos nuestra tabla de datos general.

###**Creamos la tabla HeadHunter**


In [ ]:
headhunter = pd.concat([tabla1, tabla2, tabla3], axis= 0)
headhunter.head(5)

Puesto  \
0   Client Data Analytics & Models Analyst   
1  Pricing Data Analytics & Models Analyst   
2                             Data Analyst   
3                 DATA SCIENTIST ANALYST I   
4                     Data Analyst Trainee   

                                         Descripcion              Ciudad  \
0  Client Data Analytics & Models Analyst\nCountr...  Boadilla del Monte   
1  Pricing Data Analytics & Models Analyst\nCount...  Boadilla del Monte   
2  Job Description\nEn Logicalis Spain queremos i...              Madrid   
3  Area:\nCORPORATE & INVESTMENT BANKING\nCompany...              Madrid   
4  En Alkemy España ¡Seguimos creciendo! Buscamos...               Spain   

  Nombre_Empresa  Salario_Minimo  Salario_Maximo Moneda  \
0      Santander             NaN             NaN    EUR   
1      Santander             NaN             NaN    EUR   
2      Logicalis             NaN             NaN    EUR   
3           BBVA             NaN             NaN    EUR   
4         Alkemy             NaN             NaN    EUR   

          Fecha_Publicacion  
0  2022-12-02T00:00:00.000Z  
1  2023-01-23T00:00:00.000Z  
2  2023-02-02T00:00:00.000Z  
3  2022-11-08T00:00:00.000Z  
4  2022-11-24T00:00:00.000Z

Estudiamos la composición de la tabla Headhunter

In [ ]:
headhunter.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10430 entries, 0 to 9999
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Puesto             10430 non-null  object 
 1   Descripcion        10430 non-null  object 
 2   Ciudad             10429 non-null  object 
 3   Nombre_Empresa     10430 non-null  object 
 4   Salario_Minimo     4445 non-null   float64
 5   Salario_Maximo     4308 non-null   float64
 6   Moneda             7177 non-null   object 
 7   Fecha_Publicacion  10430 non-null  object 
dtypes: float64(2), object(6)
memory usage: 733.4+ KB


Importamos la tabla HeadHunter para poder trabajar en la visualización

## Cambio de divisas

Gracias a nuestros asesores, nos dimos cuenta que el análisis de salarios no podría funcionar, ya que teníamos diferentes monedas y algunos valores eran superiores solo por el hecho de la moneda en la que estaba.



Creamos una función que busca los tipo de cambios de monedas a EUR, utilizando una API gratuita que genera los tipo de cambio y creamos otra función que ejecuta este cambio

In [ ]:
import requests

def get_exchange_rates():
    url = "https://api.exchangerate-api.com/v4/latest/EUR"
    response = requests.get(url)
    data = response.json()
    return data.get("rates", {})

def actualizar_tasas_conversion():
    tasas_conversion = {
        'USD': 0.0,
        'SGD': 0.0,
        'GBP': 0.0,
        'INR': 0.0,
        'MXN': 0.0,
        'IRR': 0.0,
        'EUR': 1.00
    }

    tasas_actualizadas = get_exchange_rates()

    for moneda in tasas_conversion.keys():
        if moneda in tasas_actualizadas:
            tasas_conversion[moneda] = tasas_actualizadas[moneda]

    return tasas_conversion

tasas_conversion = actualizar_tasas_conversion()
print(tasas_conversion)

{'USD': 1.04, 'SGD': 1.4, 'GBP': 0.828, 'INR': 90.6, 'MXN': 21.29, 'IRR': 44413.31, 'EUR': 1}


In [ ]:
def convertir_a_euros(df, Salario_Minimo, Salario_Maximo, columna_moneda):
    for index, row in df.iterrows():
        moneda = row[columna_moneda]
        if moneda in tasas_conversion:
            tasa = tasas_conversion[moneda]
            # Convertir salarios a euros
            df.at[index, Salario_Minimo] = row[Salario_Minimo] / tasa
            df.at[index, Salario_Maximo] = row[Salario_Maximo] / tasa
            # Cambiar la moneda a 'EUR'
            df.at[index, columna_moneda] = 'EUR'

    return df

headhunter_final = convertir_a_euros(headhunter, 'Salario_Minimo', 'Salario_Maximo', 'Moneda')

headhunter_final[['Salario_Minimo', 'Salario_Maximo', 'Moneda']].head


<bound method NDFrame.head of       Salario_Minimo  Salario_Maximo Moneda
0                NaN             NaN    EUR
1           40000.00        45000.00    EUR
2           27000.00        30000.00    EUR
3                NaN             NaN    EUR
4           24000.00        27000.00    EUR
...              ...             ...    ...
9995        88934.56       123668.48    EUR
9996        99840.00        99840.00    EUR
9997        80140.32       127738.00    EUR
9998        57456.88        73565.44    EUR
9999        30000.00        36000.00    EUR

[10430 rows x 3 columns]>

In [ ]:
# IMPORTAR CSV en conjunto.
headhunter_final.to_csv('headhunter.csv', index=False)



# Preguntas Archivo

In [ ]:
# Leer el archivo CSV
headhunter = pd.read_csv('/content/drive/MyDrive/HEAD HUNTERS/Data Sets (NO BORRAR NI EXPORTAR)/headhunter.csv')

##¿Cuantas ofertas hay en el data set?

In [ ]:
no_ofertas = headhunter.shape[0]
print(f"El data set contiene {no_ofertas} ofertas")

El data set contiene 10430 ofertas


##¿Cual es el salario medio?

In [ ]:
salario_medio = headhunter['Salario_Maximo'].mean()
print(f"El salario medio es de {salario_medio}")

El salario medio es de 24758.69989503356


##¿Cuales son los trabajos mejor pagados y donde?

In [ ]:
trabajo_mejor_pagado = headhunter.sort_values(by='Salario_Maximo', ascending=False)
trabajo_mejor_pagado.head(3)

trabajo_mejor_pagado[['Puesto', 'Salario_Maximo','Ciudad']].head(3)

Puesto  Salario_Maximo  \
1510            CISO - Principal Security Architect        394617.6   
4591  Site Reliability Engineering Manager - Staked        372000.0   
2681  Site Reliability Engineering Manager - Staked        372000.0   

          Ciudad  
1510  Austin, TX  
4591   Miami, FL  
2681  Boston, MA

##¿Tienen los mismos ofertantes ambas fuentes de datos?

 No, ya que los 3 archivos no tienen el mismo origen, pese a tener una gran similitud en los datos. Uno de los CSV proviene de la empresa GLASSDOOR y los 2 restantes provienen de la base de datos de la empresa que nos ha contratado.


# EDA Tabla HeadHunter

Leemos los archivos

In [ ]:
# Leer el archivo CSV
headhunter = pd.read_csv('/content/headhunter (3).csv')

In [ ]:
import seaborn as sns
import matplotlib.pyplot as plt

##**Estudiamos la tabla HeadHunter**

##1.⁠ ⁠Tipo de datos # REVISAR

In [ ]:
print(headhunter.dtypes)

print('------- IS NULL ------')

print(headhunter.isnull().sum())

Puesto                object
Descripcion           object
Ciudad                object
Nombre_Empresa        object
Salario_Minimo       float64
Salario_Maximo       float64
Moneda                object
Fecha_Publicacion     object
dtype: object
------- IS NULL ------
Puesto                0
Descripcion           0
Ciudad                0
Nombre_Empresa        0
Salario_Minimo        0
Salario_Maximo        0
Moneda               89
Fecha_Publicacion     0
dtype: int64


En la tabla conjunta aparecen 89 nulos en las columa Moneda

Podemos identificar los siguentes tipos de datos: </br> </br>
  **(Objet)**</br>

*  Puesto
*  Descripción
*  Ciudad
*  Nombre Empresa
*  Moneda
*  Fecha Publicación</br>

**(float64)**

*   Salario Minimo
*   Salario Maximo

Es importante notar que aunque la columna Fecha_Publicacion está clasificada como tipo objeto, probablemente contiene fechas </br> </br>
Los datos de tipo objeto (texto) son útiles para información categórica o descriptiva, mientras que los datos numéricos permiten realizar cálculos y análisis cuantitativos como los salarios ofrecidos en las ofertas de trabajo.









**Estudiamos si hay nulos en las tablas uno por uno**

Hemos observado que en las tablas individuales no aparece ningun nulo en ninguna fila

In [ ]:
# Tabla GlassDoor

print('----- TABLA GLASSDOOR -----')

tabla1_1.info()

#  Ofertas de empleo sample

print('----- OFERTAS DE EMPLEO SIMPLE  -----')

tabla2.info()
#O fertas de empleo

print('----- OFERTAS DE EMPLEO  -----')

tabla3.info()




----- TABLA GLASSDOOR -----


NameError: name 'tabla1_1' is not defined

##2.⁠ ⁠Ofertas por Ciudad


Madrid y Barcelona dominan el mercado laboral con una cantidad significativamente mayor de ofertas en comparación con otras ciudades.
Hay una caída pronunciada en el número de ofertas después de las primeras 4 ciudades (Madrid, Barcelona, Londres y Singapur), lo que indica una alta concentración de oportunidades en estas áreas.
La mayoría de las ciudades con más ofertas están ubicadas en Europa, destacando especialmente España
Las ciudades en la parte inferior del gráfico tienen una cantidad mucho menor de ofertas, lo que podría indicar mercados laborales más pequeños o especializados.

In [ ]:
# Contar las ofertas de trabajo por ciudad
offers_by_city = headhunter['Ciudad'].value_counts().reset_index()
offers_by_city.columns = ['Ciudad', 'Ofertas']

# Mostrar las primeras filas del resultado
print(offers_by_city.head())

In [ ]:
# Contar las ofertas de trabajo por ciudad
offers_by_city = headhunter['Ciudad'].value_counts().reset_index()
offers_by_city.columns = ['Ciudad', 'Ofertas']

# Seleccionar las 30 ciudades con más ofertas de trabajo
top_30_cities = offers_by_city.head(30)

# Generar un gráfico de puntos para las 30 ciudades con más ofertas
plt.figure(figsize=(12, 8))
plt.scatter(top_30_cities['Ofertas'], top_30_cities['Ciudad'], color='blue', alpha=0.7, s=100)
plt.title('Distribución de las 30 Ciudades con Más Ofertas de Trabajo', fontsize=16)
plt.xlabel('Número de Ofertas', fontsize=12)
plt.ylabel('Ciudad', fontsize=12)
plt.grid(axis='x', linestyle='--', alpha=0.7)
plt.tight_layout()
plt.show()

##3.Tipos de trabajo con mayor frecuencia

Los 5 trabajos que aparecen con mayor frecuencia son: </br>
Project Manager (22 ofertas) </br>
Software Engineer (14 ofertas) </br>
Business Analyst (13 ofertas) </br>
Data Analyst (12 ofertas) </br>
Product Manager (10 ofertas) </br>
Estos resultados muestran una clara demanda en el mercado laboral para roles de gestión de proyectos, desarrollo de software y análisis de datos.
</br> La presencia de Project Manager en la cima de la lista sugiere una fuerte necesidad de profesionales capaces de liderar y coordinar equipos en diversos proyectos.

In [ ]:
import pandas as pd

# Obtener el top 5 de trabajos más frecuentes
top_jobs = headhunter['Puesto'].value_counts().head(5)

# Crear una tabla con los resultados
top_jobs_table = top_jobs.reset_index()
top_jobs_table.columns = ['Puesto', 'Frecuencia']

# Mostrar la tabla
print(top_jobs_table)

In [ ]:
import matplotlib.pyplot as plt

# Obtener el top 5 de trabajos más frecuentes
top_jobs = headhunter['Puesto'].value_counts().head(5)

# Crear una tabla con los resultados
top_jobs_table = top_jobs.reset_index()
top_jobs_table.columns = ['Puesto', 'Frecuencia']

# Generar un gráfico de barras horizontal y colorido
plt.figure(figsize=(10, 6))
colors = ['#FF5733', '#33FF57', '#3357FF', '#F1C40F', '#8E44AD']
plt.barh(top_jobs_table['Puesto'], top_jobs_table['Frecuencia'], color=colors)
plt.title('Top 5 Trabajos Más Frecuentes', fontsize=16)
plt.xlabel('Frecuencia', fontsize=12)
plt.ylabel('Puesto', fontsize=12)
plt.gca().invert_yaxis()  # Invertir el eje Y para que el puesto más frecuente esté arriba
plt.tight_layout()
plt.show()


##4.⁠ Cuantas ofertas hay por año.


In [ ]:

headhunter['Fecha_Publicacion'] = pd.to_datetime(headhunter['Fecha_Publicacion'], errors='coerce')

headhunter['Año'] = headhunter['Fecha_Publicacion'].dt.year
offers_per_year = headhunter['Año'].value_counts().reset_index()
offers_per_year.columns = ['Año', 'Ofertas']
mean_offers_per_year = offers_per_year['Ofertas'].mean()

print(f"La media de ofertas por año es: {mean_offers_per_year:.2f}")

In [ ]:

headhunter['Fecha_Publicacion'] = pd.to_datetime(headhunter['Fecha_Publicacion'], errors='coerce')

headhunter['Año'] = headhunter['Fecha_Publicacion'].dt.year

offers_per_year = headhunter['Año'].value_counts().sort_index()
plt.figure(figsize=(8, 8))
plt.pie(
    offers_per_year.values,
    labels=offers_per_year.index,
    autopct='%1.1f%%',
    startangle=140,
    colors=["#FF9999", "#66B2FF", "#99FF99", "#FFCC99", "#C299FF"]
)
plt.title("Distribución de Ofertas por Año")
plt.show()


Distribución de Ofertas: </br>
2021: 3.3% </br>
2022: 56.7% </br>
2023: 40.0% </br>
Podemos observar, teniendo en cuenta los datos ofrecidos, que el año que cuenta con más ofertas publicadas fue 2022 con un 56.7%.</br>
Los años 2021 tiene la media de ofertas publicada mas baja. </br>
**Visualización:** </br>
El gráfico de pastel muestra claramente que el año 2022 tiene la mayor proporción de ofertas, mientras que los otros años tienen participaciones menores.

##5⁠.⁠Conteo de ofertas por Empresas


El gráfico de barras generado muestra las 10 empresas con más ofertas de trabajo. </br>
Las empresas lideres estan foramdas por Michael Page con 139 ofertas de trabajo, seguida de Turner & Townsend con 136.</br>
TD SYNNEX, Accenture y Apple también destacan con más de 100 ofertas cada una.
</br></br>
Las primeras 5 empresas concentran una gran parte de las ofertas, lo que indica que estas compañías tienen una alta demanda de talento.
</br></br>
Estas empresas probablemente operan en sectores como consultoría, tecnología, y servicios profesionales, donde la demanda de talento / rotación es alta.

In [ ]:

offers_by_company = headhunter['Nombre_Empresa'].value_counts()

plt.figure(figsize=(10, 6))
offers_by_company.head(10).plot(kind='bar', color='skyblue')
plt.title('Top 10 Empresas con Más Ofertas de Trabajo', fontsize=16)
plt.xlabel('Empresa', fontsize=12)
plt.ylabel('Número de Ofertas', fontsize=12)
plt.xticks(rotation=45, ha='right')
plt.tight_layout()
plt.show()

#**WEB SCRAPPING**

## Ejecución web Scrapping

Comenzamos nuestra investigación sobre Web Scraping a principios de diciembre, inicialmente asumiendo que sería un proceso más sencillo. Aplicamos distintos enfoques basándonos en los conocimientos adquiridos en clases y con el apoyo de Silvia para encontrar el driver adecuado. Durante aproximadamente un mes, hasta Navidad, realizamos múltiples intentos tanto de forma individual como en las reuniones semanales de los miércoles. Sin embargo, el 8 de enero tomamos la decisión de replantear nuestra estrategia y explorar nuevas soluciones más allá de los enfoques tradicionales de Web Scraping.

A partir de ese momento, reiniciamos el proceso desde cero. Consultando con contactos del sector, recibimos la sugerencia de utilizar una VPN con una IP alemana, dado que las restricciones de Web Scraping en Alemania son más flexibles en comparación con otros países. No obstante, cabe destacar que, a nivel global, estas restricciones han ido en aumento.

Para implementar esta estrategia, investigamos cómo configurar una VPN con IP alemana desde España y determinamos que la opción más sencilla era utilizar un servicio que ofreciera servidores en Alemania. Tras una evaluación de diversas alternativas, optamos por ProtonVPN, ya que ofrecía una versión gratuita y cumplía con nuestros requerimientos. Luego de descargar la aplicación y registrarnos, verificamos que la conexión estaba activa mediante whatismyipaddress.com.



In [ ]:
!pip install selenium
!pip install selenium webdriver-manager openpyxl
!pip install fake_useragent
!wget https://dl.google.com/linux/direct/google-chrome-stable_current_amd64.deb
!dpkg -i google-chrome-stable_current_amd64.deb
!apt-get -f install



--2025-02-13 18:37:40--  https://dl.google.com/linux/direct/google-chrome-stable_current_amd64.deb
Resolving dl.google.com (dl.google.com)... 108.177.121.91, 108.177.121.190, 108.177.121.93, ...
Connecting to dl.google.com (dl.google.com)|108.177.121.91|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 114042240 (109M) [application/x-debian-package]
Saving to: ‘google-chrome-stable_current_amd64.deb.1’

google-chrome-stabl 100%[===================>] 108.76M   255MB/s    in 0.4s    

2025-02-13 18:37:40 (255 MB/s) - ‘google-chrome-stable_current_amd64.deb.1’ saved [114042240/114042240]

(Reading database ... 125081 files and directories currently installed.)
Preparing to unpack google-chrome-stable_current_amd64.deb ...
Unpacking google-chrome-stable (133.0.6943.98-1) over (133.0.6943.98-1) ...
Setting up google-chrome-stable (133.0.6943.98-1) ...
Processing triggers for mailcap (3.70+nmu1ubuntu1) ...
Processing triggers for man-db (2.10.2-1) ...
Reading package 

In [ ]:
!which google-chrome

/usr/bin/google-chrome


In [ ]:
# Importamos librerias necessarias

from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.common.keys import Keys
import time
from bs4 import BeautifulSoup
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.chrome.options import Options
from fake_useragent import UserAgent
from selenium.webdriver.support.ui import Select
import pandas as pd
import requests
import datetime
from bs4 import BeautifulSoup as bs
from selenium import webdriver
from selenium.webdriver.chrome.service import Service
from selenium.webdriver.chrome.options import Options
from selenium.webdriver.common.by import By
from selenium.webdriver.common.keys import Keys
from webdriver_manager.chrome import ChromeDriverManager
from openpyxl import Workbook

Con la VPN en funcionamiento, retomamos el Web Scraping con la expectativa de que las restricciones fueran menores. Sin embargo, al ejecutar nuestro código inicial, el driver generaba errores y no lograba localizar Chrome. Tras una serie de pruebas y búsquedas, descubrimos que, aunque Chrome estaba instalado en nuestros equipos, no venía preinstalado en Google Colab. La solución fue instalar Chrome en Colab y utilizar su ruta específica para configurar el driver correctamente.



In [ ]:
# Configuración Selenium i Chrome

chrome_options = Options()
chrome_options.add_argument('--headless')
chrome_options.add_argument('--no-sandbox')
chrome_options.add_argument('--disable-dev-shm-usage')
chrome_options.binary_location = "/usr/bin/google-chrome"
chrome_options.add_argument("--disable-blink-features=AutomationControlled")


Una vez configurado el driver, pudimos continuar con el desarrollo del código basándonos en nuestras referencias previas. Iniciamos la creación de un archivo de Excel para almacenar los datos y comenzamos la extracción de información desde la URL de Indeed. Para evitar posibles fallos debido a la latencia de la conexión a través de la VPN, incrementamos los tiempos de espera en nuestro código.


In [ ]:
# Iniciación WebDriver

service = Service(ChromeDriverManager().install())
driver = webdriver.Chrome(service=service, options=chrome_options)


Durante este proceso, nos enfrentamos a varios desafíos:

1. Filtros en las búsquedas: Indeed no permite realizar búsquedas generales sin aplicar filtros, por lo que seleccionamos la categoría Data Analyst, ya que se alineaba con nuestro proyecto.



In [ ]:
# Creación libro de Excel
wb = Workbook()
ws = wb.active
ws.title = "Data"
ws.append(['jobTitle', 'jobDetails', 'jobLocation', 'employerName', 'salary/min', 'salary/max', 'salary/currency', 'datePosted'])

Links = []


In [ ]:
# URL de búsqueda de Indeed
url =  'https://de.indeed.com/jobs?q=Data+Analyst&start=0'

driver.get(url)

time.sleep(7)

2. Interrupciones por cookies: La plataforma requería la aceptación de cookies antes de acceder al contenido. Al notar que este mensaje también aparecía al ingresar manualmente, implementamos un código para aceptar automáticamente las cookies.


In [ ]:
# Acceptación de cookies
try:
    driver.find_element(By.ID, 'onetrust-accept-btn-handler').click()
except:
    pass

link, ti, content = [], [], []


3. Incompatibilidad del código con la estructura HTML: Al ejecutar nuestro código de extracción, identificamos discrepancias con la estructura HTML de Indeed, lo que impedía la correcta navegación entre páginas. Para solucionar este problema, realizamos un análisis manual del código fuente y ajustamos nuestro script para localizar referencias clave, como el botón de "next page".


In [ ]:
# Bucle para recorrer varias páginas de resultados
for g in range(5):
    try:
        soup = bs(driver.page_source, 'html.parser')
        data = soup.findAll(class_='slider_container css-nqpl5t eu4oa1w0')

        for d in data:
            try:
                title = d.find(class_='jcs-JobTitle css-1baag51 eu4oa1w0').find('span')['title']
                ti.append(title)
                link = "https://de.indeed.com/viewjob?jk=" + d.find(class_='jcs-JobTitle css-1baag51 eu4oa1w0')['data-jk']
                Link.append(link)
                content = d.find(class_='css-1afmp4o e37uo190').text.strip()
                content.append(content)
            except Exception as e:
                print(f"Error al extraer datos: {e}")
                continue


        # Ir a la siguiente página
        try:
            next_page = driver.find_element(By.CSS_SELECTOR, 'a[data-testid="pagination-page-next"]')
            next_page.click()
            time.sleep(random.randint(5, 10))
        except Exception as e:
            print(f"No se encontró botón de siguiente página: {e}")
            break
    except Exception as e:
        print(f"Error en la página {g}: {e}")
        driver.quit()
        if driver is None:
            print("No se pudo reiniciar el WebDriver. Saliendo...")
            break



No se encontró botón de siguiente página: Message: no such element: Unable to locate element: {"method":"css selector","selector":"a[data-testid="pagination-page-next"]"}
  (Session info: chrome=133.0.6943.98); For documentation on this error, please visit: https://www.selenium.dev/documentation/webdriver/troubleshooting/errors#no-such-element-exception
Stacktrace:
#0 0x559192e31bba <unknown>
#1 0x5591928cf790 <unknown>
#2 0x559192920c80 <unknown>
#3 0x559192920e01 <unknown>
#4 0x55919296f944 <unknown>
#5 0x559192946a7d <unknown>
#6 0x55919296cccc <unknown>
#7 0x559192946823 <unknown>
#8 0x559192912a88 <unknown>
#9 0x559192913bf1 <unknown>
#10 0x559192dfb15b <unknown>
#11 0x559192dff0e2 <unknown>
#12 0x559192de801c <unknown>
#13 0x559192dffcd4 <unknown>
#14 0x559192dcc48f <unknown>
#15 0x559192e204f8 <unknown>
#16 0x559192e206c9 <unknown>
#17 0x559192e30a36 <unknown>
#18 0x7bdae24c3ac3 <unknown>



<ipython-input-78-57d02642c0ad>:5: DeprecationWarning: Call to deprecated method findAll. (Replaced by find_all) -- Deprecated since version 4.0.0.
  data = soup.findAll(class_='slider_container css-nqpl5t eu4oa1w0')


In [ ]:
# Extraer información de cada oferta de trabajo
for link, comp, tt in zip(link, content, ti):
    driver.get(link)
    time.sleep(2)
    soup = bs(driver.page_source, 'html.parser')

    data = soup.findAll(class_='slider_container css-nqpl5t eu4oa1w0')

    role = tt
    location = soup.find('div', attrs={'data-testid': 'jobsearch-JobInfoHeader-companyLocation'})
    location = location.text if location else "N/A"

    description = soup.find(id='jobDescriptionText')
    description = description.text if description else "N/A"

    date_posted = str(datetime.date.today())

    apply_link = "N/A"
    try:
        apply_link = soup.find(id='applyButtonLinkContainer').find('button')['href']
    except:
        pass

    Anstellungsart = "N/A"
    try:
        Anstellungsart = driver.find_element(By.CSS_SELECTOR, "div[aria-label='Anstellungsart']").text
    except:
        pass

    company_rating = "N/A"
    try:
        stars = soup.findAll('span')
        for st in stars:
            if "von 5 Sternen" in st.text.strip():
                company_rating = st.text.strip()
                break
    except:
        pass

Finalmente, tras adaptar el código a la estructura específica de Indeed, procedimos con la recolección de datos. Extrajimos la información de cada variable, la organizamos en tablas y guardamos el archivo en Colab. Como resultado, logramos descargar un archivo de Excel con 200 registros de datos extraídos con éxito.


In [ ]:
# Agregar datos al archivo Excel
ws.append(['role', 'description', 'location', 'date_posted', 'Anstellungsart ', 'apply_link', 'company_rating'])

# Guardar el archivo Excel en Google Colab
filename = "indeed_Data.xlsx"
wb.save(filename)

# Descargar el archivo Excel
from google.colab import files
files.download(filename)

# Cerrar el driver
driver.quit()

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

## Analisis Web Scrapping

**Tipo de contrato**


Se analizaron 210 ofertas laborales del portal Indeed hasta diciembre, enfocadas en el mercado europeo. A partir de estos datos se destacan las siguientes conclusiones clave:

**Trabajo Remoto**

Un 26,19% de las ofertas permiten teletrabajo, lo que muestra una clara tendencia hacia la flexibilidad laboral. Es una oportunidad interesante para reclutadores que buscan talento sin barreras geográficas.

**Trabajo Remoto**



*   Analista de Datos
*   Business Intelligence
*   Data Science </br>
Se observa mayor interés en perfiles analíticos que en cientificos.






**Empresas que más contratan**



*   Syns Work(Especializada en TI y datos)
*   Talents (Perfiles tecnologicos senior y directivos)
*   The Formula (Consultora del sector tecnológico)






**Ubicación geográfica clave**





Alemania destaca como país con mayor número de ofertas.</br>

Ciudades como Regensburg alcanzan una valoración de 5 estrellas, junto a otras ciudades alemanas con puntuaciones notables (4 estrellas).</br>


**Oportunidades detectadas**

Alta demanda de perfiles con experiencia</br>


Poca oferta para juniors, lo que abre una posibilidad para captar talento joven</br>


El sector tecnológico sigue en crecimiento, con Europa como centro estratégico</br>


## Conclusiones


**Dominio geográfico del mercado**

Europa concentra más del 80 % de las ofertas laborales, destacando especialmente ciudades como Madrid y Barcelona.


**Alta demanda en sectores tecnológicos y de consultoría**

Los perfiles más buscados están relacionados con gestión de proyectos, análisis de datos y desarrollo de software. Las empresas líderes en contratación pertenecen a estos sectores.


**Concentración empresarial**

Michael Page, Turner & Townsend y TD SYNNEX concentran una parte importante de las ofertas. Las empresas reclutadoras tienen un gran peso, aunque se ha analizado aparte su impacto real.

**Condiciones laborales**

La mayoría de las ofertas son de jornada completa. Solo un 26 % ofrece opciones de teletrabajo, lo que indica un margen de mejora en la flexibilidad laboral.

**Salarios poco transparentes**

Más de 6.000 ofertas no especifican salarios. No obstante, el salario mínimo promedio es de 45.503 € y el máximo de 63.345 €, ya convertidos a euros para homogeneizar los datos.

**Tendencia estacional clara**

El pico de contratación se da en los meses de Junio, Julio y Agosto.


**Oportunidades detectadas**

Hay escasez de ofertas para perfiles junior, lo que representa una oportunidad para captar talento joven. Alemania destaca como mercado fuerte en el ámbito tecnológico Europeo.